devices/uartnmeagps.py

```diff
@@ -0,0 +1,142 @@
+    except Exception as err:
+      # We raise here because if GPS fails, we're probably getting unuseful data entirely.
+      logging.error("Error setting up UART GPS.  Is this sensor correctly installed and the cable attached tightly:  " + str(err));
+      raise err
+
+  # Close the port when we shut down.
+  def __del__(self):
+    if self.stream:
+      self.stream.close()
+
+  # Look that keeps latitude and longitude up-to-date.
+  # TODO:  We no longer auto-set system time from GPS time.  We should re-add that maybe.
+  # TODO:  There is other data here too.  Maybe also save that.  hMSL in particular seems interesting.
+  def _read_gps_data(self):
+    while True:
+      try:
+        (raw_data, parsed_data) = self.nmea.read()
+        self.has_read_data = True
+
+        if hasattr(parsed_data, 'lon') and parsed_data.lon:
+          self.longitude = parsed_data.lon
+          self.last_good_reading = time.time()
+        if hasattr(parsed_data, 'lat') and parsed_data.lat:
+          self.latitude = parsed_data.lat
+          self.last_good_reading = time.time()
+        if hasattr(parsed_data, 'date') and hasattr(parsed_data, 'time') and parsed_data.date and parsed_data.time:
+          # We can set the system time using the date (with YYYY-MM-DD) and time (HH:MM:SS)
+          self.gpsdate = parsed_data.date
+          self.gpstime = parsed_data.time
+
+          if not self.has_set_time:
+            if self.interval:
+              epoch_seconds = None
+              epoch_seconds = calendar.timegm((self.gpsdate.year, self.gpsdate.month, self.gpsdate.day, self.gpstime.hour, self.gpstime.minute, self.gpstime.second))
+
+              if abs(time.time() - epoch_seconds) > self.interval:
+                logging.warning('Setting system clock to ' + datetime.datetime.fromtimestamp(epoch_seconds).isoformat() +
+                                ' because difference of ' + str(abs(time.time() - epoch_seconds)) +
+                                ' exceeds interval time of ' + str(self.interval))
+                os.system('date --utc -s %s' % datetime.datetime.fromtimestamp(calendar.timegm(epoch_seconds)).isoformat())
+                self.timesource.set_time(datetime.datetime.now())
+                self.has_set_time = True
+      except Exception as err:
+        if str(err) != self.last_error:
+          logging.error("UART GPS Error (duplicates will be suppressed): {}".format(str(err)))
+          self.last_error = str(err)
+
+  def publish(self):
+    logging.info('Publishing GPS data to remote')
+
+    result = False
+    try:
+      if not self.has_transmitted_device_info:
+        result = self._try_write_to_remote('GPS', 'Model', 'Generic UART NMEA GPS')
+        self.has_transmitted_device_info = True
+
+      if self.latitude and self.longitude and abs(self.latitude) <= 90 and abs(self.longitude) <= 180:
+        result = self._try_write_to_remote('GPS', 'latitude_degrees', self.latitude) or result
```

Comment:
I confess I'm a little uncomfortable that latitude and longitude are written in separate lines. The idea that one could succeed and the other fail for some reason makes me itch. Like, it's fine, but I'm not stoked about it.

Reply:
Unfortunately this is baked into my backend API architecture.  Each measurement is written independently, attached to its specific timestamp.  The good news is, if one succeeded and the other failed, that would be immediately clear to anyone inspecting or using the data.